src/main/java/backend/control/RepoOpControl.java

```diff
@@ -10,23 +12,43 @@
 import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
+import java.util.Optional;
 import java.util.concurrent.*;
 
 
 /**
- * A means of repo-level synchronisation for select RepoIO operations.
+ * A means of repo-level synchronisation for select RepoIO operations. Only one instance of this class
+ * is available at any time availabel through {@code getRepoOpControl}. A new instance can be created with
+ * {@code createRepoOpControl} and will replace any existing instance
  */
-public class RepoOpControl {
+public final class RepoOpControl {
 
     private static final Logger logger = LogManager.getLogger(RepoOpControl.class.getName());
 
     private final RepoIO repoIO;
+    private final MultiModel models;
 
     private final ExecutorService pool = Executors.newCachedThreadPool();
     private final Map<String, BlockingQueue<RepoOp>> queues = new HashMap<>();
 
-    public RepoOpControl(RepoIO repoIO) {
+    private static Optional<RepoOpControl> repoOpControlOptional = Optional.empty();
+
+    private RepoOpControl(RepoIO repoIO, MultiModel models) {
         this.repoIO = repoIO;
+        this.models = models;
+    }
+
+    /**
+     * Gets an RepoOpControl instance that has been created previously by {@code createRepoOpControl}
+     * @return
+     */
+    public static RepoOpControl getRepoOpControl() {
```

Comment:
This is the ugly part of this that I was fighting over. `RepoOpControl` depends on both `RepoIO` and `MultiModel` and so needs to be instantiated in `Logic`. `RepoIO` needs to use the same `RepoOpControl` as Logic (in the `update...` method). And `PMD` thinks it's not a good idea to make `RepoOpControl` as a static instance in `Logic` and instantiate it in `Logic`'s constructor.

One way is to break the `RepoIO`'s `update...` method into smaller methods to be called on a higher level so that `RepoIO` does not depend on `RepoOpControl` but right now the `update` method is doing certain things that make it quite hard to do that (I will explain below).

So any idea to make this nice other than the singleton instance?